src/include/mpir_pset.h

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#ifndef MPIR_PSET_H_INCLUDED
+#define MPIR_PSET_H_INCLUDED
+
+#include "mpiimpl.h"
+
+/**
+ * @brief Process set structure
+ */
+struct MPIR_Pset {
+    char *uri;                  /* Name of the process set */
+    int size;                   /* Number of processes in the set */
+    bool is_valid;              /* True if process set is valid, false if not */
+    int *members;               /* Array of members of the process set (ranks) sorted in ascending order
```

Comment:
The global pset need be able to describe ranks that are consistent across sessions, i.e. with namespace-rank pairs such as "a-0", "a-1", "b-0", etc. They will be mapped to integer world ranks for specific session, but the mapper will be session-specific. Thus, I don't think `int *member` is sufficient.  If each namespace is always within a single namespace, then we can just add the namespace field to this struct. Otherwise, we'll need a global mapper to translate "namespace-rank" into global unique ids.

Reply:
As discussed offline, the `members` are MPI ranks an we will indeed need a mapping from PMIx Process ID to MPI rank when we want to support dynamic resources/ malleability in the future. Added an explanation as a [comment](https://github.com/pmodels/mpich/pull/6430/files#diff-a7fe0a05ef65ff6d4a4ad561c5c04946f1c9cc49ca0c4821823bb36c2ea4f4daR407) in the code.

Reply:
I agree that the implementation was limited to a single namespace only.

I struggled a bit to find a good compromise with respect to the required additional framework/ infrastructure to manage PMIx processes from different namespaces. In ParaStation MPI we do have an implementation for this (internal code review ongoing), however, this introduces *considerable* changes to `mpir_pmix` (+1500 lines of code, introducing PMIx Process Group, PMIx Connect, status sync of processes across namespaces etc. ...). This would definitely go beyond the scope of this PR here where we want to add support for PMIx psets. :)

The compromise I found was stripping down our solution to its very basics: We think that the mapping of PMIx proc id to MPI rank should be done within `mpir_pmix`  because that is where this information is obtained in the first place. Any code outside of `mpir_pmix` should be able to work with uniform MPI ranks (via `MPIR_Process.rank`) no matter how many namespaces are used and how many processes might have terminated. That is why we refrained from adding such a management to `MPIR_Pset`.

In 9dac33adf1ff450ba4717bddfb6913a98f5433d2 I've added a global `pmix_data_array_t` that contains all participating PMIx processes in a specific order. This array is used to do the mapping from PMIx proc ID to MPI rank (and vice versa) based on the index that a process has in the array. There are two mapping functions added: `pmix_id_to_mpi_rank` and `mpi_rank_to_pmix_id`. Those mapping functions are put to use throughout `mpir_pmix` - also in the Pset definition. I've also made some changes to `pmix_build_nodmap` to support nodemap generation across multiple namespaces there.

As currently still only a single namespace can be detected in `pmix_init`, most of these changes probably look like they add additional code that is not really required at the moment. And that is true. However, I hope that this additional commit helps to understand how this mapping can be done in practice.

If you want to have a look at the complete solution for dynamic process management in ParaStation MPI, please let me know. I'm happy to share that code with you.